### PR TITLE
[extractor/twitch] Storyboards for Twitch VODs

### PR DESCRIPTION
**Pull request by ftk**
<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### Description of your *pull request* and other information

</details>

Added storyboard formats to twitch VoDs, similar to youtube storyboards.
Extractor does 1 more JSON request.

Also fixed a few tests.

### CODE REVIEW

File: yt_dlp/extractor/twitch.py [hunk 1]
```diff
@@ -383,10 +437,41 @@ def _extract_info_gql(self, info, item_id):
             'was_live': True,
         }
 
+    def _extract_storyboard(self, item_id, storyboard_json_url, duration):
+        spec = list(self._download_json(storyboard_json_url, item_id, "Downloading storyboard metadata JSON") or [])
+        # sort from highest quality to lowest
+        spec.sort(key=lambda x: int_or_none(x.get('width')) or 0, reverse=True)
```
Comment (pukkandan) [inline yt_dlp/extractor/twitch.py:443]: No need to sort. Core code should handle it

File: yt_dlp/extractor/twitch.py [hunk 2]
```diff
@@ -383,10 +437,41 @@ def _extract_info_gql(self, info, item_id):
             'was_live': True,
         }
 
+    def _extract_storyboard(self, item_id, storyboard_json_url, duration):
+        spec = list(self._download_json(storyboard_json_url, item_id, "Downloading storyboard metadata JSON") or [])
```
Comment (pukkandan) [inline yt_dlp/extractor/twitch.py:441]: 1. why cast to list?
2. Use single quotes
3. Must be non-fatal
```suggestion
        spec = self._download_json(storyboard_json_url, item_id, 'Downloading storyboard metadata JSON', fatal=False) or []
```

File: yt_dlp/extractor/twitch.py [hunk 3]
```diff
@@ -403,6 +488,10 @@ def _real_extract(self, url):
                 })),
             vod_id, 'mp4', entry_protocol='m3u8_native')
 
+        # extract storyboard
+        if video.get('storyboard') is not None and info.get('duration') is not None:
```
Comment (pukkandan) [inline yt_dlp/extractor/twitch.py:492]:
```suggestion
        if video.get('storyboard') and info.get('duration'):
```
or, even better imo, you could just have:
```py
formats.extend(self._extract_storyboard(vod_id, video.get('storyboard'), info.get('duration')))
```
and handle the None case in _extract_storyboard

File: yt_dlp/extractor/twitch.py [hunk 4]
```diff
@@ -383,10 +437,41 @@ def _extract_info_gql(self, info, item_id):
...
+    def _extract_storyboard(self, item_id, storyboard_json_url, duration):
+        spec = list(self._download_json(storyboard_json_url, item_id, "Downloading storyboard metadata JSON") or [])
+        # sort from highest quality to lowest
+        spec.sort(key=lambda x: int_or_none(x.get('width')) or 0, reverse=True)
+        base = base_url(storyboard_json_url)
+        for i, s in enumerate(spec):
+            count = int_or_none(s.get('count'))
+            images = list(s.get('images') or [])
+            if len(images) == 0 or count is None:
+                continue
```
Comment (pukkandan) [inline yt_dlp/extractor/twitch.py:449]:
```suggestion
            images = s.get('images')
            if not (images and count):
                continue
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/twitch.py (hunk 1 above)
Comment (ftk) [reply yt_dlp/extractor/twitch.py:443]: Then sb1 will(might) be lower quality than sb0. I'd rather it be consistent with youtube (sb0 - highest quality, sb1 - lower, etc)

Comment (pukkandan) [reply yt_dlp/extractor/twitch.py:445]: Makes sense. Could u explain that in the comment coz the intention is not obvious

**Pull request merged**